fdb-record-layer-lucene/src/test/java/com/apple/foundationdb/record/lucene/FDBLuceneIndexFailureTest.java

```diff
@@ -146,6 +171,167 @@ void basicNonGroupedPartitionedTest(boolean useLegacyAsyncToSync) {
+        Index index = COMPLEX_PARTITIONED;
+        Tuple groupingKey = Tuple.from(1);
+        int repartitionCount = 2;
+        int mergeSegmentsPerTier = 2;
+
+        final RecordLayerPropertyStorage contextProps = RecordLayerPropertyStorage.newBuilder()
+                .addProp(LuceneRecordContextProperties.LUCENE_REPARTITION_DOCUMENT_COUNT, repartitionCount)
+                .addProp(LuceneRecordContextProperties.LUCENE_MERGE_SEGMENTS_PER_TIER, (double)mergeSegmentsPerTier)
+                .addProp(LuceneRecordContextProperties.LUCENE_USE_LEGACY_ASYNC_TO_SYNC, useLegacyAsyncToSync)
+                .build();
+
+        Consumer<FDBRecordContext> schemaSetup = context -> rebuildIndexMetaData(context, COMPLEX_DOC, index);
+        long docGroupFieldValue = groupingKey.isEmpty() ? 0L : groupingKey.getLong(0);
+
+        int transactionCount = 100;
+        int docsPerTransaction = 2;
+        // create/save documents
+        long id = 0;
+        List<Long> allIds = new ArrayList<>();
+        for (int i = 0; i < transactionCount; i++) {
+            try (FDBRecordContext context = openContext(contextProps)) {
+                schemaSetup.accept(context);
+                recordStore.getIndexDeferredMaintenanceControl().setAutoMergeDuringCommit(false);
+                long start = Instant.now().toEpochMilli();
+                for (int j = 0; j < docsPerTransaction; j++) {
+                    id++;
+                    recordStore.saveRecord(createComplexDocument(id, ENGINEER_JOKE, docGroupFieldValue, start + id));
+                    allIds.add(id);
+                }
+                commit(context);
+            }
+        }
+
+        // we haven't done any merges yet, or repartitioning, so each transaction should be one new segment
+        assertEquals(Map.of(0, transactionCount),
+                getSegmentCounts(index, groupingKey, contextProps, schemaSetup));
+
+        // Run a few merges with different failure points
+        assertThrows(LuceneConcurrency.AsyncToSyncTimeoutException.class,
+                () -> explicitMergeIndex(index, contextProps, schemaSetup, true, groupingKey, 0, 0));
+        assertThrows(LuceneConcurrency.AsyncToSyncTimeoutException.class,
+                () -> explicitMergeIndex(index, contextProps, schemaSetup, true, groupingKey, 0, 5));
+        assertThrows(LuceneConcurrency.AsyncToSyncTimeoutException.class,
+                () -> explicitMergeIndex(index, contextProps, schemaSetup, true, groupingKey, 0, 10));
+
+        // Continue with the regular merge and verify that index is still valid
+        timer.reset();
+        explicitMergeIndex(index, contextProps, schemaSetup, false, groupingKey, 0, 0);
+        final Map<Integer, Integer> segmentCounts = getSegmentCounts(index, groupingKey, contextProps, schemaSetup);
+        final int partitionSize = repartitionCount == 3 ? 9 : 10;
+        final int partitionCount;
+        if (repartitionCount == 3) {
```

Comment:
I think `repartitionCount` could be inlined and this could be simplified.

Reply:
Done.